resources/assets/js/videos/videoContainer.vue

```diff
@@ -274,6 +277,24 @@ export default {
                 shape_id: this.shapes[pendingAnnotation.shape],
                 label_id: this.selectedLabel ? this.selectedLabel.id : 0,
             });
+
+            // Check polygons
+            if (annotation.shape === 'Polygon') {
+                let feature = this.getFeatureFromAnnotation(annotation);
+                let polygonValidator = new PolygonValidator(feature);
+                if (polygonValidator.isInvalidPolygon()) {
+                    // Disallow polygons with less than three non-overlapping points
+                    Messages.danger(`Invalid shape. Polygon needs at least 3 non-overlapping vertices.`);
+                    this.removeAnnotation(tmpAnnotation);
+                    return;
+                }
+
+                // If polygon is self-intersecting, create valid polygon
+                let validPolygon = polygonValidator.getValidPolygon();
+                let twoDimCoords = validPolygon.getGeometry().getCoordinates();
+                annotation.points = [twoDimCoords[0].flat()];
+            }
+
```

Comment:
I think this rather belongs into `js/videos/components/videoScreen/drawInteractions.vue`. You can add it to [`extendPendingAnnotation()`](https://github.com/biigle/core/blob/ed7d8a58e01f84d9602eee59a974442995c11e58/resources/assets/js/videos/components/videoScreen/drawInteractions.vue#L165).

Reply:
Shall I do the same with the code in [annotationCanvas](https://github.com/biigle/core/blob/0316882b5e6a4978f3b61e6dc4f8bbec227a2b89/resources/assets/js/annotations/components/annotationCanvas.vue#L516-L529)?

Reply:
I think not. The new code is added to `handleNewFeature()` which is used in several of the annotationCanvas mixins (draw, measure, magic wand, polygon brush). So the method should best stay there and the updates be applied there, too.